---
title: "Visualising gene ontology results"
author: "Aine Fairbrother-Browne & Regina Reynolds"
date: "12/19"
output:
  html_document:
    code_folding: show
    df_print: paged
    highlight: kate
    theme: paper
    toc: yes
    toc_depth: 2
    toc_float: yes
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = T, warning = F, message= F)

```

# Loading necessary libraries

```{r}

library(ggpubr)
library(gprofiler2)
library(tidyverse)
library(readxl)

```

# Check your working directory
- Paths provided within this .rmd assume you are working within the `Hackathon19_visualise_ontology` project directory. Provided you have opened `Hackathon19_visualise_ontology` as a project this should already be the case.
- You can check by doing the following:

```{r}

getwd()

```

# Running gProfiler2
- Input genes are sourced from the latest [PD GWAS](https://www.biorxiv.org/content/10.1101/388165v3).
- We will be using two lists:
    - Nearest nominated gene
    - QTL nominated gene
```{r}

# import PD GWAS table
PD_res <-
  readxl::read_excel(path = "./data/TableS2_Detailed_summary_statistics_nominated_risk_variants.xlsx")

# Creating a named list
geneLists <- 
  PD_res %>% 
  # Select two columns we wish to turn into lists
  dplyr::select(`Nearest Gene`, `QTL Nominated Gene (nearest QTL)`) %>% 
  # Coerce columns into list format
  as.list() %>% 
  # Across each list do the following...
  lapply(., function(x){
    x %>% 
      # Remove NAs
      na.omit() %>%
      # Convert to character vector
      as.character() %>% 
      # Keep only unique genes
      unique()
  }) %>% 
  # Name the lists
  setNames(., c("nearest_gene", "QTL_nom_gene"))


# running gprofiler
gprofilerOutput <- gprofiler2::gost(query = geneLists, # if lists are named then gost can use this an input
                               organism = "hsapiens", # set the organism 
                               correction_method = c("gSCS"), # select a correction method (gSCS reccomended in gost documentation)
                               domain_scope = c("annotated"), # select whether to restrict the background set
                               sources = c("GO"), # databases would you like to search for pathways & processes (others: "KEGG","REAC","WP")
                               significant = TRUE, # when TRUE displays only significant results (p<0.05)
                               evcodes = FALSE, # when TRUE displays the IDs of the genes calling each pathway
                               ordered_query = FALSE) # when TRUE the input list is ordered in some meaningful way

```

- The background list (`domain_scope` argument) is an important choice, as it should represent the population of genes your list was sampled from. E.g. If you were to perform a differential gene expression experiment in brain samples from control and PD, your background list would be all the expressed genes you tested.
    - `gprofiler2::gost` provides 3 different options: "annotated", "known", "custom". For more details see: https://biit.cs.ut.ee/gprofiler/page/docs#statistical_domain_scope
    
# What do the results mean?
```{r}

gprofilerOutput$result %>% 
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

```

- Notice that one of the columns is `query`, and within this we find the names of our inputted lists i.e. `r gprofilerOutput$result %>% .[["query"]] %>% unique()`
- gprofiler outputs a number of columns, a few of which are not entirely self-explanatory, including:
    - precision: this is the proportion of genes in the input gene list that can be assigned this term i.e. overlap.size/query.size
    - recall: this is the proportion of genes in the input list that overlap with the term i.e. overlap.size/term.size
    - parents: GO terms belong to families of terms. This column will provide the parent terms to the GO term found enriched.
- Also, as some of you may notice there are some duplicated terms between our two lists, which are shown in the table below.
```{r}

gprofilerOutput$result %>% 
  # Filter for duplicated terms
  dplyr::filter(duplicated(term_id) | duplicated(term_id, fromLast = TRUE)) %>% 
  # Rearrange column order to have term_id and term_name at beginning
  dplyr::select(term_id,term_name, everything()) %>% 
  # Arrange rows by term_id
  dplyr::arrange(term_id) %>% 
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

```

# How to visualise this information?
- There are many ways to visualise these results, with a few examples given below.
- How do you think this information would be best displayed? **This is what we're going to do over the next few days.**

# Visualising the results as a barplot  
- Remember we queried two lists, so we will want to plot this twice
```{r, fig.height=8}

source("./R/gen_gprofiler_barplot.R")

plot_list <- 
  gprofilerOutput$result %>% 
  # Group by query
  dplyr::group_by(query) %>% 
  # Split dataframe by query groups i.e. nearest_gene and QTL_nom_gene
  dplyr::group_split() %>%
  # For each dataframe plot a barplot
  lapply(., function(x){
    gen_gprofiler_barplot(DF = x,
                          num_categories = 15,
                          plot_title = "")
  })

# ggarrange allows us to arrange multiple ggplots on the same page
ggpubr::ggarrange(plotlist = plot_list,
                  labels = c("nearest_gene", "QTL_nom_gene"),
                  nrow = 2)  


```

# Visualising the results with gprofiler's in-built visualisation

```{r}

gprofiler2::gostplot(gprofilerOutput) 

```

# Visualising as a network?
- Nodes would represent GO terms and edges the genes shared between these GO terms
- Potential packages:
    - [ggraph](https://github.com/thomasp85/ggraph)
    - [ggnetwork](https://github.com/briatte/ggnetwork)
    - [geom_net](https://github.com/sctyner/geomnet)

# Session info
- It is considered good practice to always your session info in the html output of an R markdown.
- Session info will depend on the RStudio you are using, your operating system and the packages you load. Thus, when you knit the final product of your code, it may not look like it does in this version of the html output. 
```{r session info}

sessionInfo()

```

